mh/database.py

```diff
@@ -0,0 +1,328 @@
+}
+
+
+def new_random_str(length=6):
+    return "".join(random.choice(CHARSET) for _ in range(length))
+
+
+class ServerType(object):
+    OPEN = 1
+    ROOKIE = 2
+    EXPERT = 3
+    RECRUITING = 4
+
+
+class LayerState(object):
+    JOINABLE = 0
+    EMPTY = 1
+    FULL = 2
+
+
+class City(object):
+    def __init__(self, name, parent):
+        self.name = name
+        self.parent = parent
+        self.capacity = 4
+        self.state = LayerState.EMPTY
+        self.players = set()
+
+    def get_population(self):
+        return len(self.players)
+
+    def get_capacity(self):
+        return self.capacity
+
+    def get_state(self):
+        size = self.get_population()
+        if size == 0:
+            return LayerState.EMPTY
+        elif size < self.get_capacity():
+            return LayerState.JOINABLE
+        else:
+            return LayerState.FULL
+
+
+class Gate(object):
+    def __init__(self, name, parent, city_count=40, player_capacity=100):
+        self.name = name
+        self.parent = parent
+        self.state = LayerState.EMPTY
+        self.capacity = player_capacity
+        self.cities = [
+            City("City{}".format(i), self)
+            for i in range(1, city_count+1)
+        ]
+        self.players = set()
+
+    def get_population(self):
+        return len(self.players) + sum((
+            city.get_population()
+            for city in self.cities
+        ))
+
+    def get_capacity(self):
+        return self.capacity
+
+    def get_state(self):
+        size = self.get_population()
+        if size == 0:
+            return LayerState.EMPTY
+        elif size < self.get_capacity():
+            return LayerState.JOINABLE
+        else:
+            return LayerState.FULL
+
+
+class Server(object):
+    def __init__(self, name, server_type, gate_count=40, capacity=2000,
+                 addr=None, port=None):
+        self.name = name
+        self.server_type = server_type
+        self.capacity = capacity
+        self.addr = addr
+        self.port = port
+        self.gates = [
+            Gate("City Gate{}".format(i), self)
+            for i in range(1, gate_count+1)
+        ]
+        self.players = set()
+
+    def get_population(self):
+        return len(self.players) + sum((
+            gate.get_population() for gate in self.gates
+        ))
+
+    def get_capacity(self):
+        return self.capacity
+
+
+def new_servers():
+    servers = []
+    servers.extend([
+        Server("Valor{}".format(i), ServerType.OPEN)
+        for i in range(1, 5)
+    ])
+    servers.extend([
+        Server("Beginners{}".format(i), ServerType.ROOKIE)
+        for i in range(1, 3)
+    ])
+    servers.extend([
+        Server("Veterans{}".format(i), ServerType.EXPERT)
+        for i in range(1, 3)
+    ])
+    servers.extend([
+        Server("Greed{}".format(i), ServerType.RECRUITING)
+        for i in range(1, 5)
+    ])
+    return servers
+
+
+class TempDatabase(object):
```

Comment:
This class is suppose to be the interface of the persistent database that the server would have(?). I don't see much sense in saving the city or gates as part of this database since the players connected to the server, would need to be re-loaded/re-connected if a restart would come to happen.

Reply:
At some point, we'll need to synchronise the available servers if we use a decentralised model. The OPN, LMP and FMP servers are redirecting to the same address for all available servers which won't be the case when the FMP servers will be properly implemented. However, we'd need to keep track of the "registered" servers (for instance, multiple FMP servers).

I don't know if it's the best place to store that info but that's why for the time being I'm storing it in the TempDatabase.

Reply:
Ahhh, good point. I did not thought about it that way.